File: frontend/src/utils/approvalTiers.ts
// src/utils/approvalTiers.ts
//
// Frontend MIRROR of `nirmaan_stack/services/approval_tiers.py`.
// THE PYTHON FILE IS THE AUTHORITY. This exists only so the UI can TELL the user which
// path their payment or expense is about to take -- which approvals it needs, and what
// their own click will do. `approvalTiers.test.ts` reads the Python source and fails if
// the two ever disagree.
//
// THE RULE (owner note, 11/09/2026)
//     amount < 15,000            -> AUTO.   Created `Approved`; no human sees it.
//     15,000 <= amount <= 50,000 -> L1.     One signature FINISHES it.
//     amount > 50,000            -> L1+L2.  L1 forwards, the CEO approves.
//     amount <= 0                -> NOT auto. Banded by SIZE like any other amount.
//
// ⚠️ THE EDGES ARE THE RULE. "Below 15000" / "15000 to <l2>" / "Above <l2>" -- so exactly
// 15,000 needs L1, and exactly 50,000 stops AT L1. This REVERSES the expense ruling of
// 2026-09-04 (`<= 10000`, inclusive), which still stands documented in `expenseApproval.ts`
// until that module is retired. Do not "fix" one to match the other.
//
// ⚠️ THE SIGN DOES ONE THING ONLY: IT BLOCKS AUTO-APPROVAL (owner, 2026-09-15). That is
// exactly what it did before -- both ledgers guarded auto-approval with `0 < amount` and
// the sign had no other effect. A -₹500 refund therefore takes L1; a -₹60,000 one takes
// both gates. An earlier revision forced every refund to L1+L2 and the owner reversed it.

import { parseNumber } from "./parseNumber";

/** Strictly BELOW this needs no human at all. */
export const TIER_AUTO_APPROVE_BELOW = 15000;
/** Strictly ABOVE this needs the CEO on top of L1. */
// ⚠️ ALL THREE LEDGERS NOW SHARE BOTH LINES (owner, 16 Sep 2026): auto below 15,000,
// L1 only from 15,000 to 50,000, CEO above 50,000. This REVERSES the per-ledger split
// of 15 Sep, which held the expense CEO line at 30,000 -- an expense between 30,000
// and 50,000 now FINISHES AT L1 instead of going to the CEO.
export const TIER_L2_ABOVE = 50000;
/**
 * The expense ledgers' CEO line. Kept as its own name even though it now equals
 * `TIER_L2_ABOVE`, so a future divergence is one line rather than a hunt through every
 * expense call site. Mirrors `TIER_L2_ABOVE_EXPENSES` in the Python module, which is the
 * authority; `approvalTiers.test.ts` reads the Python source and pins the two together.
 */
export const TIER_L2_ABOVE_EXPENSES = 50000;

export const APPROVAL_TIERS = {
  auto: "auto",
  l1: "l1",
  l1l2: "l1_l2",
} as const;

export type ApprovalTier = (typeof APPROVAL_TIERS)[keyof typeof APPROVAL_TIERS];

/** Short chip copy for the Tier column. */
export const TIER_LABELS: Record<ApprovalTier, string> = {
  auto: "AUTO",
  l1: "L1",
  l1_l2: "L1+L2",
};

/** What the approver's click will actually do -- the reason the Tier column exists. */
export const TIER_ACTION_HINTS: Record<ApprovalTier, string> = {
  auto: "Approved on creation — no approval needed",
  l1: "Your approval finishes this",
  l1_l2: "Your approval forwards this to the CEO",
};

/**
 * Which approvals this amount needs.
 *
 * ⚠️ `parseNumber` is load-bearing, not tidiness. `Project Expenses.amount` is a `Data`
 * column, so an amount arrives as a STRING -- and a raw string compare puts "9000" above
 * "50000", routing a ₹9,000 expense to the CEO. Anything unreadable parses to 0, which
 * falls into the `<= 0` branch: L1, and never auto. Not-auto is the property that matters --
 * guessing a tier wrong costs a signature, guessing AUTO wrong lets money out unreviewed.
 */
export const requiredTier = (
  amount: string | number | null | undefined,
  l2Above: number = TIER_L2_ABOVE
): ApprovalTier => {
  const value = parseNumber(amount);
  // The sign blocks AUTO only; magnitude decides the tier. Mirrors the Python branch
  // `return TIER_L1_L2 if value < -TIER_L2_ABOVE else TIER_L1`.
  if (value <= 0)
    return value < -l2Above ? APPROVAL_TIERS.l1l2 : APPROVAL_TIERS.l1;
  if (value < TIER_AUTO_APPROVE_BELOW) return APPROVAL_TIERS.auto;
  if (value <= l2Above) return APPROVAL_TIERS.l1;
  return APPROVAL_TIERS.l1l2;
};

/** The status a NEW record is created at. */
export const initialStatus = (
  amount: string | number | null | undefined,
  l2Above: number = TIER_L2_ABOVE
): "Approved" | "Requested" =>
  requiredTier(amount, l2Above) === APPROVAL_TIERS.auto ? "Approved" : "Requested";

/**
 * Where an L1 approval LANDS a record.
 *
 * ⚠️ The behaviour change for payments: today L1 always forwards to the CEO; under this
 * rule a 15k–50k approval is FINAL, and ~912 payments a year stop reaching them.
 */
export const statusAfterL1 = (
  amount: string | number | null | undefined,
  l2Above: number = TIER_L2_ABOVE
): "CEO Pending" | "Approved" =>
  requiredTier(amount, l2Above) === APPROVAL_TIERS.l1l2 ? "CEO Pending" : "Approved";

/** Whether this amount skips approval entirely. */
export const isAutoApproved = (
  amount: string | number | null | undefined
): boolean => requiredTier(amount) === APPROVAL_TIERS.auto;

/** Whether L2 is required on top of L1. */
export const needsCEO = (
  amount: string | number | null | undefined,
  l2Above: number = TIER_L2_ABOVE
): boolean => requiredTier(amount, l2Above) === APPROVAL_TIERS.l1l2;

// ── Who raised it (owner, 2026-09-21) — mirrors `initial_status_for_raiser` ─────────────────
//
//     raised by            15,000 - 50,000        above 50,000
//     anyone else          Requested (-> L1)      Requested (-> L1 -> CEO)
//     an L1 approver       Approved               CEO Pending   (L1 cleared)
//     the CEO              Approved               Approved      (L1 and CEO cleared)
//
// L1 = the Admin profile (`Administrator` arrives here as it, via `useUserData`); the CEO is
// `CEO_AUTHORIZED_USER`. THE SERVER DECIDES (`services/approval_raiser.py`); this only lets a
// request dialog say where the record will land before it is created.

export const RAISER_LEVELS = { l1: "l1", ceo: "ceo" } as const;
export type RaiserLevel = (typeof RAISER_LEVELS)[keyof typeof RAISER_LEVELS];

export const raiserLevelOf = (
  role: string | null | undefined,
  userId: string | null | undefined,
  ceoUser: string
): RaiserLevel | null => {
  if (userId && userId === ceoUser) return RAISER_LEVELS.ceo;
  if (userId === "Administrator" || role === "Nirmaan Admin Profile") return RAISER_LEVELS.l1;
  return null;
};

/** The status a NEW record is created at, given who raises it. `null` level = `initialStatus`. */
export const initialStatusForRaiser = (
  amount: string | number | null | undefined,
  l2Above: number = TIER_L2_ABOVE,
  level: RaiserLevel | null = null
): "Approved" | "CEO Pending" | "Requested" => {
  const tier = requiredTier(amount, l2Above);
  if (tier === APPROVAL_TIERS.auto) return "Approved";
  if (level === RAISER_LEVELS.ceo) return "Approved";
  if (level === RAISER_LEVELS.l1) return tier === APPROVAL_TIERS.l1l2 ? "CEO Pending" : "Approved";
  return "Requested";
};

/**
 * One line for a request dialog when the raiser's level changes where the record lands, else
 * null (the ordinary path needs no explanation). Never shown in the auto band.
 */
export const raiserLandingNote = (
  amount: string | number | null | undefined,
  l2Above: number = TIER_L2_ABOVE,
  level: RaiserLevel | null = null
): string | null => {
  if (!level || requiredTier(amount, l2Above) === APPROVAL_TIERS.auto) return null;
  const status = initialStatusForRaiser(amount, l2Above, level);
  if (status === "Approved") {
    return level === RAISER_LEVELS.ceo
      ? "You hold the CEO approval, so this is approved as soon as you request it."
      : "You hold L1 approval, so this is approved as soon as you request it.";
  }
  return "You hold L1 approval, so this skips L1 and goes straight to the CEO.";
};

File: nirmaan_stack/services/approval_tiers.py
# Copyright (c) 2026, Nirmaan (Stratos Infra Technologies Pvt. Ltd.) and contributors
# For license information, please see license.txt

"""Which approvals an amount needs, across all three money-out ledgers.

PURE MODULE -- no `frappe`, no database, no request context. Callable from a plain unittest
with no bench, no site and no fixtures. That is the point: this is the rule that decides
whether money needs one signature, two, or none, and it must be verifiable without standing
up an application.

THE RULE (owner note, 11/09/2026)
---------------------------------
    amount < 15,000          -> DIRECT APPROVE. Created already `Approved`; no human sees it.
    15,000 <= amount <= 50,000 -> L1 ONLY. Admin / Accountant Lead signs and that FINISHES it.
    amount > 50,000          -> L1, THEN L2 (CEO). Both are required.
    amount <= 0              -> the full path regardless of size (see the refund rule below).

⚠️ THE EDGES ARE THE RULE, NOT A DETAIL. The note reads "Below 15000", "15000 to 50000",
"Above 50000", so exactly Rs 15,000 needs L1 and exactly Rs 50,000 stops at L1. Written as
`<` and `<=` below, deliberately.

⚠️ THIS REVERSES THE EXPENSE RULE OF 2026-09-04, WHICH WAS `<= 10000` (INCLUSIVE). That ruling
made exactly Rs 10,000 auto-approve. The new lower edge is EXCLUSIVE at a different number, so
the old ruling is superseded rather than contradicted -- but anything written against "exactly
the limit auto-approves" is stale, including the note still standing in
`frontend/src/utils/expenseApproval.ts` until that module is retired.

⚠️ A REFUND / CREDIT (amount <= 0) ALWAYS TAKES THE FULL PATH. It is "less than 15,000" on a
plain reading of the number, and auto-approving it would let money move back with no review.
The lower bound is therefore `0 < amount`, not `amount < 15000` alone -- this preserves the
existing behaviour on both payments (`0 < amount < PAYMENT_AUTO_APPROVAL_THRESHOLD`) and
expenses (`0 < flt(self.amount) <= AUTO_APPROVE_LIMIT`), which both already guarded it.

WHAT THIS REPLACES
------------------
Three thresholds that are about to disagree with each other:

    api/payments/project_payments.py      PAYMENT_AUTO_APPROVAL_THRESHOLD = 10001.0
    doctype/project_expenses.py           AUTO_APPROVE_LIMIT = 10000
    doctype/non_project_expenses.py       AUTO_APPROVE_LIMIT = 10000

They are NOT deleted by this module landing -- Phase 1 adds the rule and wires nothing. The
call sites move over in Phases 4 and 5, and the constants go with them.

⚠️ `PO_REVISION_AUTO_APPROVAL_THRESHOLD = 5000` (api/po_revisions/revision_logic.py) is a
DIFFERENT THING -- the size of a PO amendment, not a payment -- and is deliberately untouched.

⚠️ AMOUNT MUST BE NUMERIC BEFORE IT REACHES HERE, AND ONE CALLER CANNOT ASSUME THAT.
`Project Expenses.amount` is a `Data` field -- Postgres `character varying` -- while the other
two ledgers store `Currency`/numeric. A raw string compare puts "9000" above "50000", so a
project expense would be routed to the CEO on the strength of its first character. `_as_number`
below coerces defensively; the call site must still not pass a raw string it never looked at.

The frontend mirror is `frontend/src/utils/approvalTiers.ts`, pinned to this file by
`test_approval_tiers.py` + `approvalTiers.test.ts`. THIS FILE IS THE AUTHORITY.
"""

from __future__ import annotations

# ── The two thresholds ──────────────────────────────────────────────────────────────────
# Strictly BELOW this needs no human at all.
TIER_AUTO_APPROVE_BELOW = 15000.0
# Strictly ABOVE this needs the CEO on top of L1.
#
# ⚠️ ALL THREE LEDGERS NOW SHARE BOTH LINES (owner, 16 Sep 2026): auto below
# 15,000, L1 only from 15,000 to 50,000, CEO above 50,000.
#
#     Project Payments       < 15,000 auto | 15,000-50,000 L1 only | > 50,000 CEO
#     Project / Non-Project  < 15,000 auto | 15,000-50,000 L1 only | > 50,000 CEO
#       Expenses
#
# ⚠️ THIS REVERSES THE PER-LEDGER SPLIT OF 15 Sep 2026, which put the expense CEO
# line at 30,000 while payments sat at 50,000. An expense between 30,000 and
# 50,000 used to need the CEO and now FINISHES AT L1 -- so anything written
# against "expenses go to the CEO above 30,000" is stale, wherever it stands.
TIER_L2_ABOVE = 50000.0
# The expense ledgers' CEO line. Deliberately KEPT as its own name even though it
# now equals the payments line: the owner has already moved this number once, and
# a named seam makes the next divergence a one-line change instead of a hunt
# through every expense call site. Every expense caller still passes it EXPLICITLY.
TIER_L2_ABOVE_EXPENSES = 50000.0

# ── The three answers ───────────────────────────────────────────────────────────────────
TIER_AUTO = "auto"
TIER_L1 = "l1"
TIER_L1_L2 = "l1_l2"

# ── The statuses this module can name ───────────────────────────────────────────────────
# Kept as plain strings rather than imported, so the module stays free of every other import.
STATUS_REQUESTED = "Requested"
STATUS_CEO_PENDING = "CEO Pending"
STATUS_APPROVED = "Approved"


def _as_number(amount) -> float:
    """Coerce to float, treating anything unreadable as 0.

    0 fails the `0 < amount` guard, so an unparseable amount takes the FULL approval path.
    That is the safe direction: the failure mode of guessing wrong here is an extra
    signature, never an unreviewed payment.
    """
    if amount is None:
        return 0.0
    try:
        return float(amount)
    except (TypeError, ValueError):
        return 0.0


def required_tier(amount, l2_above: float = TIER_L2_ABOVE) -> str:
    """`TIER_AUTO` | `TIER_L1` | `TIER_L1_L2` for this amount.

    `l2_above` is the CEO line, which differs per ledger -- see the constants.
    """
    value = _as_number(amount)
    if value <= 0:
        # ⚠️ THE SIGN DOES ONE THING ONLY: IT BLOCKS AUTO-APPROVAL (owner, 2026-09-15).
        #
        # That is exactly what it did before this module existed -- both ledgers guarded
        # auto-approval with `0 < amount`, and the sign had no other effect. Beyond that
        # a refund is banded BY ITS SIZE, like any other amount: a small one takes L1, a
        # large one takes both gates.
        #
        # ⚠️ An earlier revision of this file forced EVERY non-positive amount to L1+L2 on
        # the argument that it preserved payment behaviour (today every non-auto payment
        # reaches the CEO, because the lead gate always forwards). The owner reversed it:
        # that reasoning generalised one ledger's routing into a rule about sign, and it
        # would have sent a Rs 500 refund to the CEO.
        #
        # An unreadable amount lands here too (`_as_number` returns 0.0) and takes L1 --
        # not auto, which is the property that matters.
        return TIER_L1_L2 if value < -l2_above else TIER_L1
    if value < TIER_AUTO_APPROVE_BELOW:
        return TIER_AUTO
    if value <= l2_above:
        return TIER_L1
    return TIER_L1_L2


def initial_status(amount, l2_above: float = TIER_L2_ABOVE) -> str:
    """The status a NEW record is created at.

    `Approved` only for the auto band; everything else starts at `Requested` and waits.
    """
    return STATUS_APPROVED if required_tier(amount, l2_above) == TIER_AUTO else STATUS_REQUESTED


def status_after_l1(amount, l2_above: float = TIER_L2_ABOVE) -> str:
    """Where an L1 approval LANDS a record.

    ⚠️ THIS IS THE ONE THAT CHANGES BEHAVIOUR FOR PAYMENTS. Today every payment above the
    auto line goes `Requested -> CEO Pending -> Approved`, so L1 always forwards. Under this
    rule an L1 signature on a 15k-50k payment FINISHES the approval, and roughly 912 payments
    a year (Rs 2.65 Cr, measured) stop reaching the CEO.

    ⚠️ The bulk approve endpoint currently holds ONE target status for a whole batch
    (`_ModeConfig.approve_target_status`). It has to become per-row against this function,
    and the SR-tax-withholding gate that compares against that same constant has to move with
    it -- or a bulk-approved 15k-50k payment lands at `Approved` with no deduction, no error,
    and nothing on screen looking wrong.
    """
    return STATUS_CEO_PENDING if required_tier(amount, l2_above) == TIER_L1_L2 else STATUS_APPROVED


def is_auto_approved(amount) -> bool:
    """Whether this amount skips approval entirely. Sets `auto_approved` at the call site.

    Takes no `l2_above`: the AUTO line is 15,000 on every ledger, and the CEO line
    cannot change whether something auto-approves.
    """
    return required_tier(amount) == TIER_AUTO


def needs_ceo(amount, l2_above: float = TIER_L2_ABOVE) -> bool:
    """Whether L2 is required on top of L1."""
    return required_tier(amount, l2_above) == TIER_L1_L2
